The aim of this project is to clean the data and analyze the included used car listings. 

# Introduction

In [1]:
import numpy as np
import pandas as pd

autos = pd.read_csv(r"C:\Users\Administrator\Desktop\dataquest projects\autos.csv", encoding='Latin-1')

In [2]:
autos

dateCrawled  \
0       2016-03-24 11:52:17   
1       2016-03-24 10:58:45   
2       2016-03-14 12:52:21   
3       2016-03-17 16:54:04   
4       2016-03-31 17:25:20   
5       2016-04-04 17:36:23   
6       2016-04-01 20:48:51   
7       2016-03-21 18:54:38   
8       2016-04-04 23:42:13   
9       2016-03-17 10:53:50   
10      2016-03-26 19:54:18   
11      2016-04-07 10:06:22   
12      2016-03-15 22:49:09   
13      2016-03-21 21:37:40   
14      2016-03-21 12:57:01   
15      2016-03-11 21:39:15   
16      2016-04-01 12:46:46   
17      2016-03-20 10:25:19   
18      2016-03-23 15:48:05   
19      2016-04-01 22:55:47   
20      2016-04-01 19:56:48   
21      2016-03-27 11:38:00   
22      2016-03-23 14:52:51   
23      2016-03-12 19:43:07   
24      2016-03-13 20:40:49   
25      2016-03-18 21:44:09   
26      2016-03-10 19:38:18   
27      2016-03-07 12:51:23   
28      2016-03-09 11:56:38   
29      2016-03-08 19:55:19   
...                     ...   
371498  2016-03-11 15:50:38   
371499  2016-03-08 13:52:00   
371500  2016-03-21 23:40:49   
371501  2016-03-26 18:36:54   
371502  2016-04-01 12:59:21   
371503  2016-04-01 01:55:58   
371504  2016-03-07 17:06:35   
371505  2016-04-03 19:56:57   
371506  2016-03-20 18:47:59   
371507  2016-03-16 17:06:19   
371508  2016-04-02 16:43:27   
371509  2016-03-11 23:40:32   
371510  2016-03-06 21:11:23   
371511  2016-03-31 18:50:37   
371512  2016-03-25 18:48:06   
371513  2016-03-21 15:36:22   
371514  2016-04-04 11:45:16   
371515  2016-03-09 11:36:55   
371516  2016-04-04 09:57:12   
371517  2016-03-28 13:48:07   
371518  2016-04-02 20:37:03   
371519  2016-03-09 13:37:43   
371520  2016-03-19 19:53:49   
371521  2016-03-27 20:36:20   
371522  2016-03-21 09:50:58   
371523  2016-03-14 17:48:27   
371524  2016-03-05 19:56:21   
371525  2016-03-19 18:57:12   
371526  2016-03-20 19:41:08   
371527  2016-03-07 19:39:19   

                                                     name  seller offerType  \
0                                              Golf_3_1.6  privat   Angebot   
1                                    A5_Sportback_2.7_Tdi  privat   Angebot   
2                          Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3                                      GOLF_4_1_4__3TÜRER  privat   Angebot   
4                          Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   
5       BMW_316i___e36_Limousine___Bastlerfahrzeug__Ex...  privat   Angebot   
6                             Peugeot_206_CC_110_Platinum  privat   Angebot   
7                            VW_Derby_Bj_80__Scheunenfund  privat   Angebot   
8                    Ford_C___Max_Titanium_1_0_L_EcoBoost  privat   Angebot   
9       VW_Golf_4_5_tuerig_zu_verkaufen_mit_Anhaengerk...  privat   Angebot   
10                                      Mazda_3_1.6_Sport  privat   Angebot   
11          Volkswagen_Passat_Variant_2.0_TDI_Comfortline  privat   Angebot   
12                      VW_Passat_Facelift_35i__"7Sitzer"  privat   Angebot   
13                         VW_PASSAT_1.9_TDI_131_PS_LEDER  privat   Angebot   
14      Nissan_Navara_2.5DPF_SE4x4_Klima_Sitzheizg_Blu...  privat   Angebot   
15                           KA_Lufthansa_Edition_450€_VB  privat   Angebot   
16                                            Polo_6n_1_4  privat   Angebot   
17                            Renault_Twingo_1.2_16V_Aut.  privat   Angebot   
18                       Ford_C_MAX_2.0_TDCi_DPF_Titanium  privat   Angebot   
19                      Mercedes_Benz_A_160_Classic_Klima  privat   Angebot   
20                      Volkswagen_Scirocco_1.4_TSI_Sport  privat   Angebot   
21      BMW_530i_TÜV_7/17_Scheckheftgepflegt_sehr_gute...  privat   Angebot   
22                          Opel_Meriva_1.Hand_TÜV_3.2018  privat   Angebot   
23                                           Stadtflitzer  privat   Angebot   
24                             MERCEDES_200E__TÜV_04/2016  privat   Angebot   
25                  BMW_530d_touring_Vo

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
dateCrawled            371528 non-null object
name                   371528 non-null object
seller                 371528 non-null object
offerType              371528 non-null object
price                  371528 non-null int64
abtest                 371528 non-null object
vehicleType            333659 non-null object
yearOfRegistration     371528 non-null int64
gearbox                351319 non-null object
powerPS                371528 non-null int64
model                  351044 non-null object
kilometer              371528 non-null int64
monthOfRegistration    371528 non-null int64
fuelType               338142 non-null object
brand                  371528 non-null object
notRepairedDamage      299468 non-null object
dateCreated            371528 non-null object
nrOfPictures           371528 non-null int64
postalCode             371528 non-null int64
lastSeen              

In [4]:
autos.head()

dateCrawled                            name  seller offerType  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat   Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   

   price abtest vehicleType  yearOfRegistration    gearbox  powerPS  model  \
0    480   test         NaN                1993    manuell        0   golf   
1  18300   test       coupe                2011    manuell      190    NaN   
2   9800   test         suv                2004  automatik      163  grand   
3   1500   test  kleinwagen                2001    manuell       75   golf   
4   3600   test  kleinwagen                2008    manuell       69  fabia   

   kilometer  monthOfRegistration fuelType       brand notRepairedDamage  \
0     150000                    0   benzin  volkswagen               NaN   
1     125000                    5   diesel        audi                ja   
2     125000                    8   diesel        jeep               NaN   
3     150000                    6   benzin  volkswagen              nein   
4      90000                    7   diesel       skoda              nein   

           dateCreated  nrOfPictures  postalCode             lastSeen  
0  2016-03-24 00:00:00             0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0       66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0       90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0       91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0       60437  2016-04-06 10:17:21

# Cleaning Column Names

From the work we did in the last screen, we can make the following observations:

- The dataset contains 20 columns, most of which are strings.
- Some columns have null values, but none have more than ~20% null values.
- The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores.

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
autos.rename({"yearOfRegistration":"registration_year"}, axis=1, inplace=True)
autos.rename({"monthOfRegistration":"registration_month"}, inplace=True, axis=1)
autos.rename({"notRepairedDamage":"unrepaired_damage"}, inplace=True, axis=1)
autos.rename({"dateCreated":"ad_created"}, inplace=True, axis=1)

autos_lower = []
for c in autos.columns:
    abd = c.lower()
    autos_lower.append(abd)
    
autos.columns = autos_lower

In [7]:
autos.columns

Index(['datecrawled', 'name', 'seller', 'offertype', 'price', 'abtest',
       'vehicletype', 'registration_year', 'gearbox', 'powerps', 'model',
       'kilometer', 'registration_month', 'fueltype', 'brand',
       'unrepaired_damage', 'ad_created', 'nrofpictures', 'postalcode',
       'lastseen'],
      dtype='object')

In [8]:
autos.head()

datecrawled                            name  seller offertype  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat   Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   

   price abtest vehicletype  registration_year    gearbox  powerps  model  \
0    480   test         NaN               1993    manuell        0   golf   
1  18300   test       coupe               2011    manuell      190    NaN   
2   9800   test         suv               2004  automatik      163  grand   
3   1500   test  kleinwagen               2001    manuell       75   golf   
4   3600   test  kleinwagen               2008    manuell       69  fabia   

   kilometer  registration_month fueltype       brand unrepaired_damage  \
0     150000                   0   benzin  volkswagen               NaN   
1     125000                   5   diesel        audi                ja   
2     125000                   8   diesel        jeep               NaN   
3     150000                   6   benzin  volkswagen              nein   
4      90000                   7   diesel       skoda              nein   

            ad_created  nrofpictures  postalcode             lastseen  
0  2016-03-24 00:00:00             0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0       66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0       90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0       91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0       60437  2016-04-06 10:17:21

# Initial Exploration and Cleaning

In [9]:
autos.rename({"nrofpictures":"num_photos"}, inplace=True, axis=1)

autos.describe(include='all')

datecrawled         name  seller offertype         price  \
count                371528       371528  371528    371528  3.715280e+05   
unique               280500       233531       2         2           NaN   
top     2016-03-24 14:49:47  Ford_Fiesta  privat   Angebot           NaN   
freq                      7          657  371525    371516           NaN   
mean                    NaN          NaN     NaN       NaN  1.729514e+04   
std                     NaN          NaN     NaN       NaN  3.587954e+06   
min                     NaN          NaN     NaN       NaN  0.000000e+00   
25%                     NaN          NaN     NaN       NaN  1.150000e+03   
50%                     NaN          NaN     NaN       NaN  2.950000e+03   
75%                     NaN          NaN     NaN       NaN  7.200000e+03   
max                     NaN          NaN     NaN       NaN  2.147484e+09   

        abtest vehicletype  registration_year  gearbox        powerps   model  \
count   371528      333659      371528.000000   351319  371528.000000  351044   
unique       2           8                NaN        2            NaN     251   
top       test   limousine                NaN  manuell            NaN    golf   
freq    192585       95894                NaN   274214            NaN   30070   
mean       NaN         NaN        2004.577997      NaN     115.549477     NaN   
std        NaN         NaN          92.866598      NaN     192.139578     NaN   
min        NaN         NaN        1000.000000      NaN       0.000000     NaN   
25%        NaN         NaN        1999.000000      NaN      70.000000     NaN   
50%        NaN         NaN        2003.000000      NaN     105.000000     NaN   
75%        NaN         NaN        2008.000000      NaN     150.000000     NaN   
max        NaN         NaN        9999.000000      NaN   20000.000000     NaN   

            kilometer  registration_month fueltype       brand  \
count   371528.000000       371528.000000   338142      371528   
unique            NaN                 NaN        7          40   
top               NaN                 NaN   benzin  volkswagen   
freq              NaN                 NaN   223857       79640   
mean    125618.688228            5.734445      NaN         NaN   
std      40112.337051            3.712412      NaN         NaN   
min       5000.000000            0.000000      NaN         NaN   
25%     125000.000000            3.000000      NaN         NaN   
50%     150000.000000            6.000000      NaN         NaN   
75%     150000.000000            9.000000      NaN         NaN   
max     150000.000000           12.000000      NaN         NaN   

       unrepaired_damage           ad_created  num_photos    postalcode  \
count             299468               371528    371528.0  371528.00000   
unique                 2                  114         NaN           NaN   
top                 nein  2016-04-03 00:00:00         NaN           NaN   
freq              263182                14450         NaN           NaN   
mean                 NaN                  NaN         0.0   50820.66764   
std                  NaN                  NaN         0.0   25799.08247   
min                  NaN                  NaN         0.0    1067.00000   
25%                  NaN                  NaN         0.0   30459.00000   
50%                  NaN                  NaN         0.0   49610.00000   
75%                  NaN                  NaN         0.0   71546.00000   
max                  NaN                  NaN         0.0   99998.00000   

                   lastseen  
count                371528  
unique               182806  
top     2016-04-07 06:45:59  
freq                     17  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

We have learned that most of the columns are text value. However, the column "num_photos" looks odd since all values equal to 0.

In [10]:
autos["num_photos"].value_counts()

0    371528
Name: num_photos, dtype: int64

We will drop off this column and also "seller" and "offertype" since they mostly have only 1 value

In [11]:
autos.drop(["seller", "offertype", "num_photos"], axis=1)

datecrawled  \
0       2016-03-24 11:52:17   
1       2016-03-24 10:58:45   
2       2016-03-14 12:52:21   
3       2016-03-17 16:54:04   
4       2016-03-31 17:25:20   
5       2016-04-04 17:36:23   
6       2016-04-01 20:48:51   
7       2016-03-21 18:54:38   
8       2016-04-04 23:42:13   
9       2016-03-17 10:53:50   
10      2016-03-26 19:54:18   
11      2016-04-07 10:06:22   
12      2016-03-15 22:49:09   
13      2016-03-21 21:37:40   
14      2016-03-21 12:57:01   
15      2016-03-11 21:39:15   
16      2016-04-01 12:46:46   
17      2016-03-20 10:25:19   
18      2016-03-23 15:48:05   
19      2016-04-01 22:55:47   
20      2016-04-01 19:56:48   
21      2016-03-27 11:38:00   
22      2016-03-23 14:52:51   
23      2016-03-12 19:43:07   
24      2016-03-13 20:40:49   
25      2016-03-18 21:44:09   
26      2016-03-10 19:38:18   
27      2016-03-07 12:51:23   
28      2016-03-09 11:56:38   
29      2016-03-08 19:55:19   
...                     ...   
371498  2016-03-11 15:50:38   
371499  2016-03-08 13:52:00   
371500  2016-03-21 23:40:49   
371501  2016-03-26 18:36:54   
371502  2016-04-01 12:59:21   
371503  2016-04-01 01:55:58   
371504  2016-03-07 17:06:35   
371505  2016-04-03 19:56:57   
371506  2016-03-20 18:47:59   
371507  2016-03-16 17:06:19   
371508  2016-04-02 16:43:27   
371509  2016-03-11 23:40:32   
371510  2016-03-06 21:11:23   
371511  2016-03-31 18:50:37   
371512  2016-03-25 18:48:06   
371513  2016-03-21 15:36:22   
371514  2016-04-04 11:45:16   
371515  2016-03-09 11:36:55   
371516  2016-04-04 09:57:12   
371517  2016-03-28 13:48:07   
371518  2016-04-02 20:37:03   
371519  2016-03-09 13:37:43   
371520  2016-03-19 19:53:49   
371521  2016-03-27 20:36:20   
371522  2016-03-21 09:50:58   
371523  2016-03-14 17:48:27   
371524  2016-03-05 19:56:21   
371525  2016-03-19 18:57:12   
371526  2016-03-20 19:41:08   
371527  2016-03-07 19:39:19   

                                                     name  price   abtest  \
0                                              Golf_3_1.6    480     test   
1                                    A5_Sportback_2.7_Tdi  18300     test   
2                          Jeep_Grand_Cherokee_"Overland"   9800     test   
3                                      GOLF_4_1_4__3TÜRER   1500     test   
4                          Skoda_Fabia_1.4_TDI_PD_Classic   3600     test   
5       BMW_316i___e36_Limousine___Bastlerfahrzeug__Ex...    650     test   
6                             Peugeot_206_CC_110_Platinum   2200     test   
7                            VW_Derby_Bj_80__Scheunenfund      0     test   
8                    Ford_C___Max_Titanium_1_0_L_EcoBoost  14500  control   
9       VW_Golf_4_5_tuerig_zu_verkaufen_mit_Anhaengerk...    999     test   
10                                      Mazda_3_1.6_Sport   2000  control   
11          Volkswagen_Passat_Variant_2.0_TDI_Comfortline   2799  control   
12                      VW_Passat_Facelift_35i__"7Sitzer"    999  control   
13                         VW_PASSAT_1.9_TDI_131_PS_LEDER   2500  control   
14      Nissan_Navara_2.5DPF_SE4x4_Klima_Sitzheizg_Blu...  17999  control   
15                           KA_Lufthansa_Edition_450€_VB    450     test   
16                                            Polo_6n_1_4    300     test   
17                            Renault_Twingo_1.2_16V_Aut.   1750  control   
18                       Ford_C_MAX_2.0_TDCi_DPF_Titanium   7550     test   
19                      Mercedes_Benz_A_160_Classic_Klima   1850     test   
20                      Volkswagen_Scirocco_1.4_TSI_Sport  10400  control   
21      BMW_530i_TÜV_7/17_Scheckheftgepflegt_sehr_gute...   3699     test   
22                          Opel_Meriva_1.Hand_TÜV_3.2018   2900     test   
23                                           Stadtflitzer    450     test   
24                             MERCEDES_200E__TÜV_04/2016    500     test   
25                  BMW_530d_touring_Vollausstattung_NAVI   2500  control   
26            

# Exploring the Odometer and Price Columns

In [12]:
autos['price'].shape

(371528,)

In [13]:
autos['kilometer'].shape

(371528,)

In [14]:
autos['price'].describe()

count    3.715280e+05
mean     1.729514e+04
std      3.587954e+06
min      0.000000e+00
25%      1.150000e+03
50%      2.950000e+03
75%      7.200000e+03
max      2.147484e+09
Name: price, dtype: float64

In [15]:
autos['price'].value_counts().sort_index(ascending=False)

2147483647        1
99999999         15
99000000          1
74185296          1
32545461          1
27322222          1
14000500          1
12345678          9
11111111         10
10010011          1
10000000          8
9999999           3
3895000           1
3890000           1
2995000           1
2795000           1
1600000           2
1300000           1
1250000           2
1234566           1
1111111           2
1010010           1
1000000           5
999999           13
999990            1
911911            1
849000            1
820000            1
780000            1
745000            2
              ...  
35               18
33                1
32                1
30               55
29                2
27                1
26                1
25               33
24                1
21                1
20               51
19                3
18                3
17                5
16                2
15               27
14                5
13                7
12                8


As a research on ebay, the cheapest price for the car is 2500 and the most expensive one is sold at 2,500,000. So, we are going to remove rows where "price" is outside of that range

In [16]:
autos = autos[autos["price"].between(2500, 2500000)]

In [17]:
autos["price"].value_counts().sort_index(ascending=False)

1600000       2
1300000       1
1250000       2
1234566       1
1111111       2
1010010       1
1000000       5
999999       13
999990        1
911911        1
849000        1
820000        1
780000        1
745000        2
725000        1
700000        1
650000        1
619000        1
600000        2
599000        1
585000        1
579000        1
517895        1
500000        2
488997        1
487000        1
485000        1
466000        1
445000        1
440000        1
           ... 
2625          1
2620          1
2610          1
2601          1
2600       1717
2599        335
2598          2
2595          3
2590        174
2589          4
2580         15
2579          1
2575          2
2570          4
2569          1
2565          1
2560          4
2555         54
2550        477
2549         14
2544          1
2540          4
2530          5
2525          2
2520          2
2512          1
2511          1
2510          1
2501          1
2500       4438
Name: price, Length: 455

In [18]:
autos['kilometer'].value_counts()

150000    106805
125000     24230
100000     11304
90000       9710
80000       8990
70000       8396
60000       7750
50000       7008
40000       5994
30000       5383
20000       4684
5000        2511
10000       1662
Name: kilometer, dtype: int64

We can see that the values in this field are rounded, which might indicate that sellers had to choose from pre-set options for this field. Additionally, there are more high mileage than low mileage vehicles.

# Exploring the Date columns

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary:
- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

In [19]:
autos['datecrawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.026225
2016-03-06    0.014881
2016-03-07    0.035108
2016-03-08    0.032163
2016-03-09    0.032936
2016-03-10    0.031689
2016-03-11    0.032359
2016-03-12    0.036566
2016-03-13    0.016808
2016-03-14    0.036536
2016-03-15    0.032041
2016-03-16    0.029174
2016-03-17    0.030808
2016-03-18    0.013022
2016-03-19    0.035039
2016-03-20    0.037153
2016-03-21    0.034692
2016-03-22    0.031444
2016-03-23    0.031586
2016-03-24    0.030016
2016-03-25    0.033239
2016-03-26    0.032789
2016-03-27    0.031307
2016-03-28    0.035979
2016-03-29    0.033758
2016-03-30    0.032970
2016-03-31    0.031498
2016-04-01    0.034951
2016-04-02    0.036453
2016-04-03    0.040719
2016-04-04    0.038464
2016-04-05    0.012718
2016-04-06    0.003273
2016-04-07    0.001639
Name: datecrawled, dtype: float64

Looks like the site was crawled daily over roughly a one month period in March and April 2016. The distribution of listings crawled on each day is roughly uniform.

In [20]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2014-03-10    0.000005
2015-03-20    0.000005
2015-06-11    0.000005
2015-06-18    0.000005
2015-08-07    0.000005
2015-08-10    0.000005
2015-09-09    0.000005
2015-10-14    0.000005
2015-11-02    0.000005
2015-11-08    0.000005
2015-11-10    0.000005
2015-11-12    0.000005
2015-11-13    0.000005
2015-11-17    0.000005
2015-11-23    0.000010
2015-11-24    0.000010
2015-12-06    0.000005
2015-12-17    0.000005
2015-12-27    0.000005
2015-12-28    0.000005
2015-12-30    0.000010
2016-01-02    0.000005
2016-01-03    0.000005
2016-01-06    0.000005
2016-01-07    0.000010
2016-01-10    0.000020
2016-01-13    0.000015
2016-01-14    0.000005
2016-01-15    0.000005
2016-01-16    0.000010
                ...   
2016-03-09    0.032970
2016-03-10    0.031522
2016-03-11    0.032520
2016-03-12    0.036345
2016-03-13    0.018261
2016-03-14    0.035230
2016-03-15    0.031909
2016-03-16    0.029522
2016-03-17    0.030373
2016-03-18    0.013868
2016-03-19    0.034027
2016-03-20    0.037187
2016-03-21 

There is a large variety of ad created dates. Most fall within 1-2 months of the listing date, but a few are quite old, with the oldest at around 9 months.

In [21]:
autos['lastseen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.000959
2016-03-06    0.003096
2016-03-07    0.004041
2016-03-08    0.006257
2016-03-09    0.007553
2016-03-10    0.009377
2016-03-11    0.011202
2016-03-12    0.018907
2016-03-13    0.006892
2016-03-14    0.010938
2016-03-15    0.013804
2016-03-16    0.014078
2016-03-17    0.024679
2016-03-18    0.006173
2016-03-19    0.013912
2016-03-20    0.017204
2016-03-21    0.017307
2016-03-22    0.018212
2016-03-23    0.016211
2016-03-24    0.016857
2016-03-25    0.016534
2016-03-26    0.014029
2016-03-27    0.014793
2016-03-28    0.019562
2016-03-29    0.020775
2016-03-30    0.021269
2016-03-31    0.021680
2016-04-01    0.022013
2016-04-02    0.022879
2016-04-03    0.023407
2016-04-04    0.023094
2016-04-05    0.140984
2016-04-06    0.250422
2016-04-07    0.150900
Name: lastseen, dtype: float64

The crawler recorded the date it last saw any listing, which allows us to determine on what day a listing was removed, presumably because the car was sold.

The last three days contain a disproportionate amount of 'last seen' values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.

In [22]:
autos['registration_year'].describe()

count    204427.000000
mean       2006.592935
std          69.266831
min        1000.000000
25%        2003.000000
50%        2007.000000
75%        2010.000000
max        9999.000000
Name: registration_year, dtype: float64

The year that the car was first registered will likely indicate the age of the car. Looking at this column, we note some odd values. The minimum value is 1000, long before cars were invented and the maximum is 9999, many years into the future.

# Dealing with Incorrect Registration Year Data

As car was invented in 1885, we will remove all rows that have registration_year before 1885 and after 2020 (which is the year we are living in)

In [23]:
autos = autos[autos["registration_year"].between(1885,2020)]

In [24]:
autos['registration_year'].describe()

count    204376.000000
mean       2005.769449
std           7.341748
min        1910.000000
25%        2003.000000
50%        2007.000000
75%        2010.000000
max        2019.000000
Name: registration_year, dtype: float64

In [25]:
autos['registration_year'].value_counts(normalize=True).sort_index()

1910    0.000034
1911    0.000005
1923    0.000015
1927    0.000010
1928    0.000010
1929    0.000034
1930    0.000020
1931    0.000020
1932    0.000020
1933    0.000024
1934    0.000020
1935    0.000024
1936    0.000029
1937    0.000059
1938    0.000034
1939    0.000020
1940    0.000010
1941    0.000020
1942    0.000010
1943    0.000024
1944    0.000010
1945    0.000020
1946    0.000005
1947    0.000024
1948    0.000015
1949    0.000015
1950    0.000078
1951    0.000078
1952    0.000054
1953    0.000064
          ...   
1990    0.003743
1991    0.003885
1992    0.004369
1993    0.003533
1994    0.004002
1995    0.005558
1996    0.006312
1997    0.007741
1998    0.012462
1999    0.018559
2000    0.024881
2001    0.033942
2002    0.041698
2003    0.053813
2004    0.068384
2005    0.083415
2006    0.088998
2007    0.081477
2008    0.076335
2009    0.074368
2010    0.059464
2011    0.058383
2012    0.045539
2013    0.029646
2014    0.022899
2015    0.013167
2016    0.012981
2017    0.0203

It appears that most vehicles were registered in the past 20 years

# Exploring Price by Brand

In [26]:
print(autos['brand'].unique())

['audi' 'jeep' 'skoda' 'ford' 'volkswagen' 'nissan' 'bmw' 'opel' 'citroen'
 'honda' 'mercedes_benz' 'mini' 'smart' 'volvo' 'renault' 'hyundai'
 'alfa_romeo' 'kia' 'seat' 'subaru' 'porsche' 'mitsubishi' 'peugeot'
 'fiat' 'mazda' 'toyota' 'chevrolet' 'sonstige_autos' 'dacia' 'chrysler'
 'jaguar' 'suzuki' 'rover' 'saab' 'land_rover' 'lancia' 'trabant' 'lada'
 'daihatsu' 'daewoo']


In [27]:
autos['brand'].value_counts()

volkswagen        43170
bmw               28922
mercedes_benz     25138
audi              23252
opel              14346
ford              10073
renault            5057
peugeot            5031
skoda              4191
seat               3605
fiat               3582
mini               3295
toyota             3246
smart              2982
mazda              2757
citroen            2585
sonstige_autos     2403
hyundai            2341
nissan             2243
porsche            2112
volvo              1688
kia                1585
honda              1340
suzuki             1291
chevrolet          1291
mitsubishi         1176
alfa_romeo         1102
dacia               735
jeep                699
land_rover          687
chrysler            679
jaguar              515
subaru              330
saab                264
daihatsu            180
lancia              143
trabant             131
lada                124
rover                65
daewoo               20
Name: brand, dtype: int64

German facturers have 4 out of top 5 brands, represent almost 50% of the listings. Volkswagen is by far the most popular manufacture since its sales are almost doubled than the 2nd ranking which is bmw.

Some manufacturers have small percentage so we will limite our analysis by only keeping those who have more than 5%

In [28]:
brand_counts = autos['brand'].value_counts(normalize=True)
common_brands = brand_counts[brand_counts > 0.05].index
print(common_brands)

Index(['volkswagen', 'bmw', 'mercedes_benz', 'audi', 'opel'], dtype='object')


In [31]:
price_per_brand = {}

for b in common_brands:
    brand = autos[autos["brand"] == b]
    mean_price = brand["price"].mean()
    price_per_brand[b] = int(mean_price)
    
price_per_brand

{'volkswagen': 8663,
 'bmw': 11011,
 'mercedes_benz': 11201,
 'audi': 11952,
 'opel': 6382}

As per our analysis, BMW, Mercedes Benz and Audi are the most expensive car while Opel is the most affordable. Volkswagen is in between those 2 groups and this could explain why Volks are the most popular manufacturers since clients probably think they have the advantages of both expensive and inexpensive groups.

#   Storing Aggregate Data in a DataFrame

In [33]:
mean_price = pd.Series(price_per_brand)
pd.DataFrame(mean_price, columns=['mean_price'])

mean_price
volkswagen           8663
bmw                 11011
mercedes_benz       11201
audi                11952
opel                 6382

In [36]:
mileage_per_brand = {}

for b in common_brands:
    brand = autos[autos["brand"] == b]
    mean_mileage = brand["kilometer"].mean()
    mileage_per_brand[b] = int(mean_mileage)
    
mean_mileage = pd.Series(mileage_per_brand).sort_values(ascending=False)
mean_price = pd.Series(price_per_brand).sort_values(ascending=False)

table = pd.DataFrame(mean_mileage, columns=['mean_mileage'])

In [37]:
table['mean_price'] = mean_price

table

mean_mileage  mean_price
bmw                  129207       11011
mercedes_benz        126081       11201
audi                 124265       11952
volkswagen           118799        8663
opel                 112284        6382

The range of car mileages does not vary as much as the prices do by brand, instead all falling within 10% for the top brands. There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage.